File: 2800-minimum-string-length-after-removing-substrings/2800-minimum-string-length-after-removing-substrings.cpp
class Solution {
public:
    int minLength(string s) {
        stack<char> chars;
        for (char c: s) {
            if (!chars.empty() && (c == 'B' && chars.top() == 'A' || c == 'D' && chars.top() == 'C')) {
                chars.pop();
            } else {
                chars.push(c);
            }
        }
        return chars.size();
    }
};

File: 2800-minimum-string-length-after-removing-substrings/2800-minimum-string-length-after-removing-substrings.java
class Solution {
    public int minLength(String s) {
        char[] stacks = new char[s.length()+1];
        int last = -1;
        for (char c: s.toCharArray()) {
            if (last > -1 && (c == 'D' && stacks[last] == 'C' || c == 'B' && stacks[last] == 'A')) {
                last--;
            } else {
                last++;
                stacks[last] = c;
            }
        }
        return ++last;
    }
}